include/fair_shared_mutex.hpp

```diff
@@ -122,139 +179,187 @@ class shared_mutex_base {
             //
-            node* p = request.next;
-            while (p != &queue_ && (p->status & node_status_mask) == 1) {
-              p->status |= 2;
-              p = p->next;
+            if (RwLockFairness::phased) {
+              // PhaseFairness: mark all queued shared-lock nodes
+              for (node* p = request.next; p != &queue_; p = p->next) {
+                if ((p->status & node_status_mask) == 1) {
+                  p->status |= 2;
+                }
+              }
+            } else {
+              // TaskFairness: mark immediately subsequent shared-lock nodes group
+              node* p = request.next;
+              while (p != &queue_ && (p->status & node_status_mask) == 1) {
+                p->status |= 2;
+                p = p->next;
+              }
             }
             cv_.notify_all();
           }
           wq_erase(&request);
+          YAMC_DEBUG_DUMPQ("<<try_lockwait/false");
           return false;
         }
       }
       wq_erase(&request);
     }
     wq_push_locknode(0);
+    YAMC_DEBUG_DUMPQ("<<try_lockwait/true");
     return true;
   }
 
   void impl_lock_shared(std::unique_lock<std::mutex>& lk)
   {
+    YAMC_DEBUG_DUMPQ(">>lock_shared");
     if (!wq_shared_lockable()) {
       node request = {1, 0, 0};  // shared-lock
       wq_push_back(&request);
+      YAMC_DEBUG_DUMPQ("  lock_shared/wait");
       while (request.status != 3) {
         cv_.wait(lk);
       }
       wq_erase(&request);
     }
     wq_push_locknode(1);
+    YAMC_DEBUG_DUMPQ("<<lock_shared");
   }
 
   bool impl_try_lock_shared()
   {
+    YAMC_DEBUG_DUMPQ(">>try_lock_shared");
     if (!wq_shared_lockable()) {
+      YAMC_DEBUG_DUMPQ("<<try_lock_shared/false");
       return false;
     }
     wq_push_locknode(1);
+    YAMC_DEBUG_DUMPQ("<<try_lock_shared/true");
     return true;
   }
 
   void impl_unlock_shared()
   {
+    YAMC_DEBUG_DUMPQ(">>unlock_shared");
     assert(queue_.next == &locked_ && (locked_.status & node_status_mask) == 3);
     locked_.status -= node_nthread_inc;
     if (locked_.status < node_nthread_inc) {
       // all current shared-locks was unlocked
       wq_pop_locknode();
       cv_.notify_all();
     }
+    YAMC_DEBUG_DUMPQ("<<unlock_shared");
   }
 
   template<typename Clock, typename Duration>
   bool impl_try_lockwait_shared(std::unique_lock<std::mutex>& lk, const std::chrono::time_point<Clock, Duration>& tp)
   {
+    YAMC_DEBUG_DUMPQ(">>try_lockwait_shared");
     if (!wq_shared_lockable()) {
       node request = {1, 0, 0};  // shared-lock
       wq_push_back(&request);
+      YAMC_DEBUG_DUMPQ("  try_lockwait_shared/wait");
       while (request.status != 3) {
         if (cv_.wait_until(lk, tp) == std::cv_status::timeout) {
           if (request.status == 3)  // re-check predicate
             break;
           wq_erase(&request);
+          YAMC_DEBUG_DUMPQ("<<try_lockwait_shared/false");
           return false;
```

Comment:
`try_lock_shared_{for,until}` timeout means that other thread own exclusive-lock.
The subsequent `unlock` will call `cv_.notify_all()` unconditionally, so notification is not required here.